google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableUnaryOperationCallable.java

```diff
@@ -0,0 +1,189 @@
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.api.gax.tracing.ApiTracer;
+import com.google.api.gax.tracing.ApiTracerFactory;
+import com.google.api.gax.tracing.SpanName;
+import com.google.common.base.Preconditions;
+import io.grpc.Status;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper to convert a fake {@link ServerStreamingCallable} (ie only up to 1 response) into a {@link
+ * UnaryCallable}. It is intended to be the outermost callable of a chain.
+ *
+ * <p>Responsibilities:
+ *
+ * <ul>
+ *   <li>Operation level metrics
+ *   <li>Configuring the default call context
+ *   <li>Converting the result to a future
+ */
+class BigtableUnaryOperationCallable<ReqT, RespT> extends UnaryCallable<ReqT, RespT> {
+  private static final Logger LOGGER =
+      Logger.getLogger(BigtableUnaryOperationCallable.class.getName());
+  Logger logger = LOGGER;
+
+  private final ServerStreamingCallable<ReqT, RespT> inner;
+  private final ApiCallContext defaultCallContext;
+  private final ApiTracerFactory tracerFactory;
+  private final SpanName spanName;
+  private final boolean allowNoResponse;
+
+  public BigtableUnaryOperationCallable(
+      ServerStreamingCallable<ReqT, RespT> inner,
+      ApiCallContext defaultCallContext,
+      ApiTracerFactory tracerFactory,
+      SpanName spanName,
+      boolean allowNoResponse) {
+    this.inner = inner;
+    this.defaultCallContext = defaultCallContext;
+    this.tracerFactory = tracerFactory;
+    this.spanName = spanName;
+    this.allowNoResponse = allowNoResponse;
+  }
+
+  @Override
+  public ApiFuture<RespT> futureCall(ReqT req, ApiCallContext apiCallContext) {
+    apiCallContext = defaultCallContext.merge(apiCallContext);
+
+    ApiTracer apiTracer =
+        tracerFactory.newTracer(
+            apiCallContext.getTracer(), spanName, ApiTracerFactory.OperationType.Unary);
+
+    apiCallContext = apiCallContext.withTracer(apiTracer);
+
+    UnaryFuture f = new UnaryFuture(apiTracer, allowNoResponse);
+    inner.call(req, f, apiCallContext);
+    return f;
+  }
+
+  class UnaryFuture extends AbstractApiFuture<RespT> implements ResponseObserver<RespT> {
+    private final ApiTracer tracer;
+    private final boolean allowNoResponse;
+
+    private StreamController controller;
+    private final AtomicBoolean upstreamCancelled = new AtomicBoolean();
+    private boolean responseReceived;
+    private @Nullable RespT response;
+
+    private UnaryFuture(ApiTracer tracer, boolean allowNoResponse) {
+      this.tracer = Preconditions.checkNotNull(tracer, "tracer can't be null");
+      this.allowNoResponse = allowNoResponse;
+      this.responseReceived = false;
+    }
+
+    @Override
+    public void onStart(StreamController controller) {
+      this.controller = controller;
+      controller.disableAutoInboundFlowControl();
+      // Request 2 to detect protocol bugs
```

Comment:
why? is this necessary?

Reply:
to preserve the current behavior

Reply:
https://github.com/grpc/grpc-java/blob/dae078c0a636fd4b71ca7d20c333dd00b0b0714a/stub/src/main/java/io/grpc/stub/ClientCalls.java#L544